### PR TITLE
Increase form update resilience and track analytics for badly behaved servers that don't provide form hashes

### LINKED ISSUE #4895: Attempt to invoke virtual method 'java.lang.String org.odk.collect.forms.Form.getFormMediaPath()' on a null object reference
#### Software and hardware versions 
Collect v2021.3.0

#### Problem description
https://console.firebase.google.com/u/2/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/be730c9d272352bd5b8685ab0d197899?time=last-seven-days&versions=v2021.3.0%20(4316)&sessionEventKey=6184E33F02B600015883C970B2D53C78_1605436904284051748

```
Caused by java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.odk.collect.forms.Form.getFormMediaPath()' on a null object reference
       at org.odk.collect.android.utilities.FormUtils.getMediaFiles(FormUtils.java:20)
       at org.odk.collect.android.formmanagement.ServerFormsDetailsFetcher.areNewerMediaFilesAvailable(ServerFormsDetailsFetcher.java:130)
       at org.odk.collect.android.formmanagement.ServerFormsDetailsFetcher.fetchFormDetails(ServerFormsDetailsFetcher.java:86)
       at org.odk.collect.android.tasks.DownloadFormListTask.doInBackground(DownloadFormListTask.java:63)
       at org.odk.collect.android.tasks.DownloadFormListTask.doInBackground(DownloadFormListTask.java:39)
       at android.os.AsyncTask$3.call(AsyncTask.java:378)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run(Thread.java:929)
```

#### Steps to reproduce the problem
1. Upload a simple form with media files to Central and download it in Collect
2. Update the form bumping its version and download in Collect
3. Fill one form and save it
4. Delete the newer version of the form
5. Try to open the list of forms to download

The app doesn't crash but loads and loads and in logs I get the same NPE

If you have a form but without media files and you do the same, then there is no crash or infinitive loading
but the message that an update is available is not visible what is also a bug.

I used this form: [formTest.zip](https://github.com/getodk/collect/files/7485591/formTest.zip)

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #4895

#### What has been done to verify that this works as intended?
I added automated tests and reviewed implemented changes.

#### Why is this the best possible solution? Were any other approaches considered?
The problem seems to be on a server side so this just makes the code more resistant to wrong data provided by servers.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This does not require testing since reproducing the bug is difficult.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by grzesiek2010** [comment]
@seadowg I'm creating this draft because as you know I will be off till Monday so you might need to take over if I'm not able to finish today.

I was able to come up with another scenario for this crash which is:
1. You have a form with media files downloaded
2. There is an updated version of the form on the server (updated form version not only media files)
3. When you try do download form details and check if it's a newer version its hash is null or empty. 

The first commit is a test that covers that scenario.

Maybe there are even more scenarios... I need to spend more time on that code.

**Comment by grzesiek2010** [comment]
> its hash is null or empty

generally any wrong hash provided by a server might cause the same issue so: null, empty or just wrong

If my theory from https://github.com/getodk/collect/issues/4888#issuecomment-964070762 is right it might be related...

**Comment by seadowg** [comment]
> I was able to come up with another scenario for this crash

Right. We basically don't trust the version identifier on the server, so we compare the hashes and then the media file hashes. The new code we added for the patch (grabbing the correct form to compare media files) relies on the version though. Maybe we should be using the hash there instead as well? 

Either way I guess we just need to be careful we're not passing around a `null` form. I'd be happy with a Kotlin conversion if you fancy it? Should make catching these crashes easier - I think our current code will potentially fail to compile in Kotlin.

**Comment by grzesiek2010** [comment]
> The new code we added for the patch (grabbing the correct form to compare media files) relies on the version though. Maybe we should be using the hash there instead as well?

I agree that it would be better to relay on one parameter to reduce the risk and I implemented it. Generally yeah there are two options either server provides wrong hashes or form version. The second option is no longer possible because now I use hashes everywhere. For the case with invalid hashes we can use analytics. https://github.com/getodk/collect/pull/4906/files#diff-d241c237e50a73d5233d52f1d174193cc3692fe2aa977b654ac5e02ff39f13d7R78 
What do you think?

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.java [hunk 1]
```diff
@@ -159,16 +146,12 @@ private static boolean isMediaFileAlreadyDownloaded(List<File> localMediaFiles,
...
 
-    private Form getFormByVersion(List<Form> forms, @Nullable String expectedVersion) {
-        for (Form form : forms) {
-            if (Objects.equals(form.getVersion(), expectedVersion)) {
-                return form;
-            }
-        }
-        return null;
+    private boolean isHashValid(String hash) {
+        return hash != null && hash.startsWith("md5:");
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.java:157]: I think we need a test for the invalid (doesn't start with `md5:` case)

File: collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcherTest.java [hunk 2]
```diff
@@ -167,6 +172,22 @@ public void whenAFormExists_andItsNewerVersionWithUpdatedMediaFilesHasBeenAlread
         assertThat(form.isNotOnDevice(), is(false));
     }
 
+    @Test
+    public void whenAFormExists_andItsNewerVersionIsAvailableButHasNullHash_isNotNewOrUpdated() throws Exception {
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcherTest.java:193]: I'm just trying to think through this behaviour and if there are any problems in it being the case. Before this situation would just crash right?

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormUtils.java [hunk 3]
```diff
@@ -16,6 +18,7 @@ private FormUtils() {
 
     }
 
+    @NonNull
     public static List<File> getMediaFiles(Form form) {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FormUtils.java]: We should probably add a `@NonNull` to the param as well.

**Comment by seadowg** [comment]
> The second option is no longer possible because now I use hashes everywhere. For the case with invalid hashes we can use analytics. https://github.com/getodk/collect/pull/4906/files#diff-d241c237e50a73d5233d52f1d174193cc3692fe2aa977b654ac5e02ff39f13d7R78
What do you think?

Yeah as far I can see the behaviour is the same for `null` and invalid hashes after this change - they just show as not being updated. The only difference is that we don't go on to crash during the media file checks. I think analytics here would be good as from the number of crashes, it seems like there are a few people with servers/workflows that cause this.

File: collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcherTest.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcherTest.java:193]: Yes this test was failing with the same stacktrace like we can see in firebase reports.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormUtils.java (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/utilities/FormUtils.java]: Done.

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.java:157]: Like this https://github.com/getodk/collect/pull/4906/commits/ec965920fe9b15e78dee2867513851ec17f87d96?

**Review by seadowg — APPROVED** [review]
I think this is looking good! I did a quick Kotlin conversion to check there were no obvious nullable calls left over, and it looks fine. I think it'd be good to do some Kotlin conversion and take a look at restructuring the tests a little (now there's more of them), but that's a follow-up I can take care of.

Before merging:
* @kkrawczyk123 if you have a chance before the end of the day could you look at doing a quick pass on this (form downloading, checking for updates etc). As @grzesiek2010 says, there's not much to reproduce without a broken server, but I'd still feel better having this put through its paces with a working server given how many edge cases keep coming up. I'd say a PR like this only needs testing on one Android version.
* @grzesiek2010 are you able to quickly rework the history so we have a rebase rather than a merge commit? If you don't have time I can take care of that.

**Comment by grzesiek2010** [comment]
> @grzesiek2010 are you able to quickly rework the history so we have a rebase rather than a merge commit? If you don't have time I can take care of that.

Done.

**Comment by kkrawczyk123** [comment]
**Tested with success!**
Verified on Androids: 5.1, 8.1 and 9.0.

Tested cases:
- form update: Aggregate, Central, Google Drive
- check for form update periodically
- automatic download
- forms with version and media
- match exactly 
- fill different form versions
- send different form versions to Central
- tried scenario from issue

**Pull request merged**